pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -134,6 +142,89 @@ func (service serviceMetadata) match(rule *rbacproto.AccessRule) bool {
 	return true
 }
 
+// attributesEnforcedInDynamicMetadataMatcher returns true if the given attribute should be enforced
+// via dynamic metadata matcher. This means these attributes are depending on the output of authn filter.
+func attributesEnforcedInDynamicMetadataMatcher(k string) bool {
+	switch k {
+	case attrSrcNamespace, attrSrcUser, attrSrcPrincipal, attrRequestPrincipal, attrRequestAudiences,
+		attrRequestPresenter:
+		return true
+	}
+	return strings.HasPrefix(k, attrRequestClaims)
+}
+
+func generateMetadataStringMatcher(keys []string, v *metadata.StringMatcher) *metadata.MetadataMatcher {
+	paths := make([]*metadata.MetadataMatcher_PathSegment, 0)
+	for _, k := range keys {
+		paths = append(paths, &metadata.MetadataMatcher_PathSegment{
+			Segment: &metadata.MetadataMatcher_PathSegment_Key{Key: k},
+		})
+	}
+	return &metadata.MetadataMatcher{
+		Filter: authn.AuthnFilterName,
+		Path:   paths,
+		Value: &metadata.MetadataMatcher_Value{
+			MatchPattern: &metadata.MetadataMatcher_Value_StringMatch{
+				StringMatch: v,
+			},
+		},
+	}
+}
+
+// createDynamicMetadataMatcher creates a MetadataMatcher for the given key, value pair.
+func createDynamicMetadataMatcher(k, v string) *metadata.MetadataMatcher {
+	var keys []string
+	forceRegexPattern := false
+
+	if k == attrSrcNamespace {
+		// Proxy doesn't have attrSrcNamespace directly, but the information is encoded in attrSrcPrincipal
+		// with format: cluster.local/ns/{NAMESPACE}/sa/{SERVICE-ACCOUNT}, so we change the key to
+		// attrSrcPrincipal and change the value to a regular expression to match the namespace part.
+		keys = []string{attrSrcPrincipal}
+		v = fmt.Sprintf(`*/ns/%s/*`, v)
+		forceRegexPattern = true
```

Comment:
Instead of doing regex matching for namespace, can we extract namespace information from principal in authn filter. This way, we only need to do exact string matching here. I think this will make the logic simpler and more efficient. What do you think?

Reply:
proxy today already understands identity format ("spiffe://..."). I think it is ok to let proxy extract namespace info from principal. Pilot does not need to understand the identity/principal format. It simply consumes "namespace" attribute.

Reply:
It only stripped the "spiffe://" prefix (if found), but has no knowledge about the following contents.

Reply:
Correct. My point is, Istio has to understand the identity format in one of the places, proxy or pilot. If we put the logic in proxy, the code will be simpler.